applications/FluidDynamicsApplication/custom_elements/two_fluid_navier_stokes.cpp

```diff
@@ -164,11 +158,80 @@ void TwoFluidNavierStokes<TElementData>::CalculateLocalSystem(
                     ComputeGaussPointEnrichmentContributions(data, Vtot, Htot, Kee_tot, rhs_ee_tot);
                 }
 
+                Matrix int_shape_function, int_shape_function_enr_neg, int_shape_function_enr_pos;
+                GeometryType::ShapeFunctionsGradientsType int_shape_derivatives;
+                Vector int_gauss_pts_weights;
+                std::vector< array_1d<double,3> > int_normals_neg;
+
+                if (rCurrentProcessInfo[SURFACE_TENSION] || rCurrentProcessInfo[MOMENTUM_CORRECTION]){
+                    ComputeSplitInterface(
+                        data,
+                        int_shape_function,
+                        int_shape_function_enr_pos,
+                        int_shape_function_enr_neg,
+                        int_shape_derivatives,
+                        int_gauss_pts_weights,
+                        int_normals_neg,
+                        p_modified_sh_func);
+                }
+
+                if (rCurrentProcessInfo[MOMENTUM_CORRECTION]){
+                    MatrixType lhs_acc_correction = ZeroMatrix(NumNodes*NumNodes,NumNodes*NumNodes);
+
+                    double positive_density;
+                    double negative_density;
+
+                    GeometryType::Pointer p_geom = this->pGetGeometry();
+
+                    for (unsigned int intgp = 0; intgp < int_gauss_pts_weights.size(); ++intgp){
+                        double u_dot_n = 0.0;
+                        for (unsigned int i = 0; i < NumNodes; i++){
+                            u_dot_n += int_shape_function(intgp,i)*(*p_geom)[i].GetValue(DISTANCE_DIFFERENCE);
+                            /* for (unsigned int dim = 0; dim < NumNodes-1; dim++){
+                                u_dot_n += int_shape_function(intgp,i)*data.Velocity(i,dim)*(int_normals_neg[intgp])[dim];
+                            } */
+
+                            if (data.Distance[i] > 0.0){
+                                positive_density = data.NodalDensity[i];
+                            } else {
+                                negative_density = data.NodalDensity[i];
+                            }
+                        }
+
+                        u_dot_n /= data.DeltaTime;
+
+                        for (unsigned int i = 0; i < NumNodes; i++){
+                            for (unsigned int j = 0; j < NumNodes; j++){
+                                for (unsigned int dim = 0; dim < NumNodes-1; dim++){
+                                    lhs_acc_correction( i*(NumNodes) + dim, j*(NumNodes) + dim) +=
+                                        int_shape_function(intgp,i)*int_shape_function(intgp,j)*u_dot_n*int_gauss_pts_weights(intgp);
+                                }
+                            }
+                        }
+                    }
+
+                    lhs_acc_correction = (negative_density - positive_density)*lhs_acc_correction;
+                    noalias(rLeftHandSideMatrix) += lhs_acc_correction;
```

Comment:
For the sake of efficiency I'd avoid the creation of the auxiliary `lhs_acc_correction` and `tempU` arrays. Instead I'd directly assemble in `rLeftHandSideMatrix` and `rRightHandSideVector`.

Reply:
I need to subtract the residual of the last iteration from RHS. The most straightforward way is to have an explicit LHS matrix. The other option is to have nested loops.

Reply:
My point is that you already have the nested loops to fill the auxiliary LHS. You can avoid declaring these if you directly assemble in the `rLeftHandSideMatrix` and `rRightHandSideVector`.

Reply:
I can do that by accessing rRightHandSideVector[i_node] and rLeftHandSideMatrix(i_node, j_node) in the nested loops, however, future changes/debugs would be very hard for others. I do not think this is a good idea since this creation of the new matrix/vector is done only at the cut (let's say a very small portion of the elements).

Reply:
OK. Let's leave it as it is for the moment. We can enhance the performance in a future PR.